pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1230,25 +1262,70 @@ func updatePipelineRunStatusFromTaskRuns(logger *zap.SugaredLogger, pr *v1beta1.
 			// Since this was recovered now, add it to the map, or it might be overwritten
 			taskRunByPipelineTask[pipelineTaskName] = taskrun.Name
 		}
+		if _, ok := childRefByPipelineTask[pipelineTaskName]; minimalEmbedded && !ok {
+			// This taskrun was missing from the status.
+			// Add it without conditions, which are handled in the next loop
+			logger.Infof("Found a TaskRun %s that was missing from the PipelineRun status", taskrun.Name)
+
+			// Since this was recovered now, add it to the map, or it might be overwritten
+			childRefByPipelineTask[pipelineTaskName] = &v1beta1.ChildStatusReference{
+				TypeMeta: runtime.TypeMeta{
+					APIVersion: v1beta1.SchemeGroupVersion.String(),
+					Kind:       "TaskRun",
+				},
+				Name:             taskrun.Name,
+				PipelineTaskName: pipelineTaskName,
+			}
+		}
 	}
 	// Then loop by pipelinetask name over all the TaskRuns associated to Conditions
 	for pipelineTaskName, actualConditionTaskRuns := range conditionTaskRuns {
-		taskRunName, ok := taskRunByPipelineTask[pipelineTaskName]
-		if !ok {
+		ok := false
+		taskRunName := ""
+
+		if taskRunName, ok = taskRunByPipelineTask[pipelineTaskName]; !ok && fullEmbedded {
 			// The pipelineTask associated to the conditions was not found in the pipelinerun
 			// status. This means that the conditions were orphaned, and never added to the
 			// status. In this case we need to generate a new TaskRun name, that will be used
 			// to run the TaskRun if the conditions are passed.
-			taskRunName = resources.GetTaskRunName(pr.Status.TaskRuns, pipelineTaskName, pr.Name)
+			taskRunName = resources.GetTaskRunName(pr.Status.TaskRuns, pr.Status.ChildReferences, pipelineTaskName, pr.Name)
 			pr.Status.TaskRuns[taskRunName] = &v1beta1.PipelineRunTaskRunStatus{
 				PipelineTaskName: pipelineTaskName,
 				Status:           nil,
 				ConditionChecks:  nil,
 			}
 		}
+
+		if !ok && minimalEmbedded {
```

Comment:
this can be `if minimalEmbedded` since `ok` here will always be false.

Reply:
I'm not sure - I know I got bit by not checking `ok` as well somewhere, but looking at the code, I think this is probably not the case. I'll change it and see if anything blows up. =)

Reply:
ah wait my bad I was thinking in the case where you keep the `if ok := ...` then the scope of `ok` would be the if block, but that's not the case here. But I still think this logic could use some changes-- if you have flag = minimal, there's no reason to check the status.TaskRuns, and if you have flag = both, childRefs should be used as the source of truth.

Reply:
The logic is right, I think, with the `!ok` removed - we still need to be adding to both `TaskRuns` and `ChildReferences` for `both`. But we should just be generating the `taskRunName` once.

Reply:
I'm simplifying it, though. =)